constraint-solver/src/indexed_constraint_system.rs

```diff
@@ -280,6 +281,68 @@ impl<T: RuntimeConstant + Substitutable<V>, V: Clone + Hash + Ord + Eq>
     }
 }
 
+impl<T: RuntimeConstant + Substitutable<V> + Hash, V: Clone + Hash + Ord + Eq>
+    IndexedConstraintSystem<T, V>
+{
+    pub fn apply_expression_substitutions(
+        &mut self,
+        substitutions: impl IntoIterator<Item = (GroupedExpression<T, V>, GroupedExpression<T, V>)>,
+    ) {
+        for (expression, substitution) in substitutions {
+            if let Some(var) = expression.try_to_simple_unknown() {
+                self.substitute_by_unknown(&var, &substitution);
+            } else {
+                self.substitute_expression_in_bus_interactions(&expression, &substitution);
+            }
+        }
+    }
+}
+
+impl<T: RuntimeConstant + Hash, V: Clone + Hash + Ord + Eq> IndexedConstraintSystem<T, V> {
+    pub fn substitute_expression_in_bus_interactions(
+        &mut self,
+        lhs: &GroupedExpression<T, V>,
+        substitution: &GroupedExpression<T, V>,
+    ) {
+        let substitution_vars = substitution
+            .referenced_unknown_variables()
+            .unique()
+            .collect_vec();
+        let items = lhs
+            .referenced_unknown_variables()
+            .map(|v| BTreeSet::from_iter(&self.variable_occurrences[v]))
+            .reduce(|a, b| a.intersection(&b).cloned().collect())
+            .unwrap_or_default()
+            .into_iter()
+            .cloned()
+            .collect_vec();
+        let lhs_hash = hash(lhs);
+        for item in &items {
+            if let ConstraintSystemItem::BusInteraction(i) = item {
+                // We only substitute in bus interactions, as the lhs is a bus interaction.
+                let bus_interaction = &mut self.constraint_system.bus_interactions[*i];
+                for expr in bus_interaction.fields_mut() {
+                    if hash(expr) == lhs_hash && expr == lhs {
```

Comment:
Not sure any more if hashing is really faster here.

Reply:
Yeah, would be surprised, because `==` can stop as soon as there is a difference?